### ISSUE TITLE
Add a collection of short example programs to the project

### ISSUE THREAD

**Issue by jcoliz** [ShapeCrawler/ShapeCrawler#803]
I propose a top-level project folder named (e.g.) 'examples'. Under this would be numerous very small example programs each of which demonstrate a single API, in the spirit of a unit test. Each would output a presentation which we could then load into PowerPoint to confirm the API worked visually.

The problem this would solve is: Today when creating a new ShapeCrawler feature, there are three test modalities needed:
1. Create a presentation in PowerPoint, then load it in ShapeCrawler to ensure the getter gets the expected value, and the setter can modify it. We have a unit test pattern for this.
2. Round trip through the constructor/getter/setter. Again, unit tests are great for this.
3. Create a presentation in ShapeCrawler then load it in PowerPoint to ensure it looks right. Can't do this in unit tests. 😒

To cover the last one, I usually create a branch in the consuming app. But it would be better to have an isolated test case. Then also YOU have to create your own app to test the feature when the PR comes. It would be better if the PR comes with an example, and you can just run that and load it into PowerPoint yourself to confirm.

This could also be added to the unit tests to at least build and run the examples as one more safety net. Further, you could attach the outputs to the CI pipeline, so it would be easy to inspect whether a problem had cropped up in an API.

Finally, this serves a documentation purpose as well to help users in their quest to understand how to use the library.

**Comment by ashahabov** [comment]
@jcoliz, I am still thinking about the suggestion. In the meantime, here are some thoughts:

**Folder `examples`.** We already had an `examples` folder where I sometimes wrote small tests. Later, I moved them to the [Wiki](https://github.com/ShapeCrawler/ShapeCrawler/wiki/Examples). _In my opinion_, it is easier for users to search for examples on the Wiki page than to look through folders. Actually, I am not happy with the current quantity and quality of the documentation. I am working on moving the documentation to [GitBook](https://shape.gitbook.io/shapecrawler). Any help or suggestions are welcome!

**Maintainer Testing.** I do manual testing when I review a Pull Request: I clone the contributor’s fork and check out its branch, or I do it after merging the Pull Request. In both situations, I call `IPresentation.SaveAs()` in the existing test to see the output in PowerPoint. But in a test like the one below, yes, I cannot do this. I must create a new temporary unit test for my manual testing. Maybe this is what you mean?


```C#
[Test]
[SlideShape("autoshape-grouping.pptx", 1, "TextBox 7", "SimSun")]
public void EastAsianName_Getter_returns_font_for_East_Asian_characters(IShape shape, string expectedFontName)
{
    // Arrange
    var autoShape =  shape;
    var font = autoShape.TextBox!.Paragraphs[0].Portions[0].Font;

    // Act
    var fontName = font.EastAsianName;

    // Assert
    fontName.Should().Be(expectedFontName);
}
```

**Comment by jcoliz** [comment]
> It is easier for users to search for examples on the Wiki page than to look through folders

Yes, 100%. Documentation is a secondary benefit. Mostly I am after validation.

> I moved them to the [Wiki](https://github.com/ShapeCrawler/ShapeCrawler/wiki/Examples). 

I recommend having BOTH. Putting the examples into the CI build ensures that they stay working. From there, selected examples can be copied to documentation.

The big problem is that documentation writing is hard and gets out of date. Adding some running code to exercise new features has to be done _anyway_, we're just giving people a place to put it.

> I am working on moving the documentation to [GitBook](https://shape.gitbook.io/shapecrawler). Any help or suggestions are welcome!

Commendable! Honestly, I can't think of any individually-maintained project with ShapeCrawler's scope that has exemplary docs. At this point I'm pretty well used to just digging into the unit tests with a new library for anything more than cursory usage.

> I clone the contributor’s fork and check out its branch, or I do it after merging the Pull Request. In both situations, I call `IPresentation.SaveAs()` in the existing test to see the output in PowerPoint. 

Ok, given that workflow, here is what I suggest:

1. Keep the `IPresentation.SaveAs()` in the test.
2. Allow test writer to describe how to validate the file in the UI
3. Add the resulting file to the test context
4. Hopefully this would result in the files being directly available in the build results. (Azure Pipelines does this). If not, adjust the CI steps to copy the files into the build results.

Thus, if a contributor saves the output of a test, you don't have to do any of that. You just jump over to the build results and grab the file whose name corresponds to the test and open it up.

>  I must create a new temporary unit test for my manual testing. Maybe this is what you mean?

This would suffice. What I actually mean is a stand-alone compilation unit. Here is what I do on my end before submitting a change:

```Powershell
cd ShapeCrawler.Examples
dotnet new console -o ComponentName.FeatureName
dotnet sln add ComponentName.FeatureName
cd ComponentName.FeatureName
dotnet add reference ../../ShapeCrawler/src/ShapeCrawler
git add .
git commit -a -m "build: Add ComponentName.FeatureName"
```

And then edit `Program.cs` to create a presentation, add the new items, and save it.

Allowing SaveAs in the unit tests would suffice. Separate examples has the slight benefit that you can be a bit more expansive than you'd want to be in a single unit test. e.g. Here is a snipped from my properties tester:

```c#
foreach(var p in props.GetType().GetProperties())
{
    if (p.PropertyType == typeof(string))
    {
        p.SetValue(props,$">{p.Name}");
    }
}
```

**Comment by ashahabov** [comment]
I think I understand your idea to let the Maintainer and Contributors have access to "pptxTestOutput" without even needing to run code locally from the `examples` folder.

There are many definitions of the difference between Unit and Integration tests and what to call a Unit in a Unit test. But I believe that a test is a unit test as long as its workflow does not involve external environments (services), such as a database or a Web API. In my opinion, the Hard Drive is also an external service. So, when we save files in the test on the hard drive, we turn our unit test into an integration test. That means we are testing the integration with the hard drive, which does not make sense for us. We also should not forget about performance issues caused by I/O.

Regarding CI integration, we might have a separate Git Workflow that runs in parallel and does not affect the main [test.yml](https://github.com/ShapeCrawler/ShapeCrawler/blob/master/.github/workflows/test.yml). You mentioned Azure Pipelines. Maybe it is more flexible and has a more convenient UI for accessing pptxTestOutput. Unfortunately, I do not have much experience with Azure Pipelines, so I am not sure how to set it up, and I am open to ideas on that.

So, from the above, I would suggest saving presentations only in the planned `examples` tests.
